src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs

```diff
@@ -92,18 +92,25 @@ public async Task<string> SetupAsync(SetupContext context)
             var initialState = context.ShellSettings.State;
             try
             {
+                // Set the shell state to 'Initializing' to guarantee that any subsequent HTTP requests
+                // will result in a 'Service Unavailable' response while the tenant setup is in progress.
+                context.ShellSettings.AsInitializing();
+
                 var executionId = await SetupInternalAsync(context);
 
-                if (context.Errors.Any())
+                if (context.Errors.Count > 0)
                 {
                     context.ShellSettings.State = initialState;
                     await _shellHost.ReloadShellContextAsync(context.ShellSettings, eventSource: false);
                 }
 
                 return executionId;
             }
-            catch
+            catch (Exception e)
             {
+                _logger.LogError(e, "An error occurred while setting up the tenant.");
+                context.Errors.Add(String.Empty, S["An error occurred while setting up the tenants: {0}", e.Message]);
```

Comment:
If we throw again, we don't need these 2 lines
We don't need to add an error to the setup context
And on throwing, the exception will be logged.